### ISSUE TITLE
Optimize Matting Laplacian to eliminate memory bottleneck

### ISSUE THREAD

**Issue by eskutcheon** [eskutcheon/mcapst#6]
Currently, I'm only able to run a training run with batch size = 3 at most with image sizes of (512, 512). It's always the call to `laplacian.coalesce()` where it runs out of CUDA memory. Since I removed all the dense tensor options, I can remove this and do some general cleanup.

To optimize it further, I'm planning to try the following:
1. Remove any calls to `_postprocess` and leave `laplacian` as a list of `torch.sparse.Tensor` objects for the loss function.
	- This was unnecessary since I can't do batched multiplication using `torch.sparse.mm` anyway, since it requires 2D matrices.
2. Reuse the same `indices` retrieved from the cache for each batch index when constructing the Laplacian. Also, directly save the flattened indices in the cache with the repeated first dimension.
	- This was another aspect that was really just a leftover when I was still trying to write a nice, reusable tensor cache. There was no reason to expand the indices that way to begin with.
3. Compute the content image batch's Laplacian within a `torch.no_grad` context manager.
	- I didn't worry about doing this before because autograd should implicitly only trace the use of the stylized images during back-propagation, but to be safe, I should add this.
4. (maybe) Replace some Einstein summation operations with calls to `opt_einsum` which seems highly optimized for large matrix multiplications with a higher number of dimensions.
5. Try to optimize the loop in `MLLossFn.forward` and see if I could checkpoint the gradient better.

The first two options alone should immediately improve GPU memory allocation and runtime.
